File: packages/marketdata/src/candles.ts
/**
 * OHLCV candle aggregator for ONE interval (e.g. 1m, 5m).
 *
 * For each trade with (price, qty, ts):
 *   bucketStart = floor(ts / intervalMs) * intervalMs
 *
 *   If bucketStart > current.bucketStart:
 *     - "Close" the current candle (emit a `close` event, push to history).
 *     - For any intermediate buckets with no trades, emit FLAT candles
 *       (O=H=L=C = prior close, V=0). This keeps the chart continuous so
 *       the time axis doesn't compress over idle periods.
 *     - Start a fresh current candle: O=H=L=C=price, V=qty.
 *   Else if bucketStart === current.bucketStart:
 *     - Update H, L, C, V on the live candle.
 *   Else (rare, out-of-order):
 *     - Ignore. (We don't try to re-mutate already-closed candles.)
 *
 * We always emit an `update` event for the live candle so consumers
 * (chart) can mutate the rightmost bar in place each tick.
 */

export interface Candle {
  bucketStart: number;
  open: string;
  high: string;
  low: string;
  close: string;
  volume: string;
  trades: number;
}

export type CandleEventKind = 'update' | 'close';

export interface CandleEvent {
  kind: CandleEventKind;
  interval: string;     // e.g. '1m', '5m'
  candle: Candle;
}

interface InternalCandle {
  bucketStart: number;
  open: bigint;
  high: bigint;
  low: bigint;
  close: bigint;
  volume: bigint;
  trades: number;
}

export class CandleAggregator {
  private current: InternalCandle | null = null;
  private readonly history: InternalCandle[] = [];

  constructor(
    public readonly interval: string,
    public readonly intervalMs: number,
    public readonly capacity: number = 500
  ) {}

  /** Returns events to publish (one `update` for the new live candle, plus any `close`s for rolled buckets). */
  onTrade(price: bigint, qty: bigint, ts: number): CandleEvent[] {
    const bucketStart = Math.floor(ts / this.intervalMs) * this.intervalMs;
    const events: CandleEvent[] = [];

    if (this.current === null) {
      this.current = {
        bucketStart, open: price, high: price, low: price, close: price,
        volume: qty, trades: 1,
      };
      events.push(this.makeEvent('update', this.current));
      return events;
    }

    if (bucketStart > this.current.bucketStart) {
      events.push(this.makeEvent('close', this.current));
      this.archive(this.current);

      // Backfill flat candles for any empty buckets in between.
      let nextStart = this.current.bucketStart + this.intervalMs;
      const carry = this.current.close;
      while (nextStart < bucketStart) {
        const flat: InternalCandle = {
          bucketStart: nextStart, open: carry, high: carry, low: carry, close: carry,
          volume: 0n, trades: 0,
        };
        events.push(this.makeEvent('close', flat));
        this.archive(flat);
        nextStart += this.intervalMs;
      }

      this.current = {
        bucketStart, open: price, high: price, low: price, close: price,
        volume: qty, trades: 1,
      };
      events.push(this.makeEvent('update', this.current));
    } else if (bucketStart === this.current.bucketStart) {
      if (price > this.current.high) this.current.high = price;
      if (price < this.current.low) this.current.low = price;
      this.current.close = price;
      this.current.volume += qty;
      this.current.trades += 1;
      events.push(this.makeEvent('update', this.current));
    }
    // out-of-order: drop silently.
    return events;
  }

  /**
   * Close out any buckets that have ended due to wall-clock time having
   * advanced past their boundary, even without a new trade. Call this on
   * a periodic timer so the chart's live bar doesn't appear stale when
   * trading goes quiet.
   */
  tick(now: number): CandleEvent[] {
    if (this.current === null) return [];
    const bucketStart = Math.floor(now / this.intervalMs) * this.intervalMs;
    if (bucketStart <= this.current.bucketStart) return [];
    // Synthesize a flat trade at the carry price to roll the bucket forward.
    return this.onTrade(this.current.close, 0n, bucketStart);
  }

  clear(): void {
    this.current = null;
    this.history.length = 0;
  }

  /** Last N (closed + live) candles, oldest first. */
  recent(limit: number): Candle[] {
    const out: Candle[] = [];
    const start = Math.max(0, this.history.length - limit + (this.current ? 0 : 1));
    for (let i = start; i < this.history.length; i++) out.push(toPublic(this.history[i]!));
    if (this.current) out.push(toPublic(this.current));
    return out.slice(-limit);
  }

  private archive(c: InternalCandle): void {
    this.history.push(c);
    if (this.history.length > this.capacity) this.history.shift();
  }

  private makeEvent(kind: CandleEventKind, c: InternalCandle): CandleEvent {
    return { kind, interval: this.interval, candle: toPublic(c) };
  }
}

function toPublic(c: InternalCandle): Candle {
  return {
    bucketStart: c.bucketStart,
    open: c.open.toString(),
    high: c.high.toString(),
    low: c.low.toString(),
    close: c.close.toString(),
    volume: c.volume.toString(),
    trades: c.trades,
  };
}

File: packages-go/marketdata/candles.go
package main

// Candle holds OHLCV in integer base units; the `,string` tags emit them as
// decimal strings on the wire (matching the TS format) while the aggregator
// does plain int64 arithmetic — one struct, no internal/public duplication.
type Candle struct {
	BucketStart int64 `json:"bucketStart"`
	Open        int64 `json:"open,string"`
	High        int64 `json:"high,string"`
	Low         int64 `json:"low,string"`
	Close       int64 `json:"close,string"`
	Volume      int64 `json:"volume,string"`
	Trades      int   `json:"trades"`
}

type CandleEventKind string

const (
	CandleEventUpdate CandleEventKind = "update"
	CandleEventClose  CandleEventKind = "close"
)

type CandleEvent struct {
	Kind     CandleEventKind `json:"kind"`
	Interval string          `json:"interval"`
	Candle   Candle          `json:"candle"`
}

// CandleAggregator is OHLCV candle aggregator for ONE interval (e.g. 1m, 5m).
//
// For each trade with (price, qty, ts):
//
//	bucketStart = floor(ts / intervalMs) * intervalMs
//
//	If bucketStart > current.bucketStart:
//	  - "Close" the current candle (emit a `close` event, push to history).
//	  - For any intermediate buckets with no trades, emit FLAT candles
//	    (O=H=L=C = prior close, V=0). This keeps the chart continuous so
//	    the time axis doesn't compress over idle periods.
//	  - Start a fresh current candle: O=H=L=C=price, V=qty.
//	Else if bucketStart == current.bucketStart:
//	  - Update H, L, C, V on the live candle.
//	Else (rare, out-of-order):
//	  - Ignore. (We don't try to re-mutate already-closed candles.)
//
// We always emit an `update` event for the live candle so consumers
// (chart) can mutate the rightmost bar in place each tick.
type CandleAggregator struct {
	Interval   string
	IntervalMs int64
	Capacity   int

	current *Candle
	history []*Candle
}

func NewCandleAggregator(interval string, intervalMs int64, capacity int) *CandleAggregator {
	if capacity == 0 {
		capacity = 500
	}
	return &CandleAggregator{Interval: interval, IntervalMs: intervalMs, Capacity: capacity}
}

func (a *CandleAggregator) OnTrade(price, qty, ts int64) []CandleEvent {
	bucketStart := (ts / a.IntervalMs) * a.IntervalMs
	var events []CandleEvent

	if a.current == nil {
		a.current = &Candle{
			BucketStart: bucketStart, Open: price, High: price, Low: price, Close: price,
			Volume: qty, Trades: 1,
		}
		events = append(events, a.makeEvent(CandleEventUpdate, a.current))
		return events
	}

	switch {
	case bucketStart > a.current.BucketStart:
		events = append(events, a.makeEvent(CandleEventClose, a.current))
		a.archive(a.current)

		// Backfill flat candles for any empty buckets in between.
		nextStart := a.current.BucketStart + a.IntervalMs
		carry := a.current.Close
		for nextStart < bucketStart {
			flat := &Candle{
				BucketStart: nextStart, Open: carry, High: carry, Low: carry, Close: carry,
				Volume: 0, Trades: 0,
			}
			events = append(events, a.makeEvent(CandleEventClose, flat))
			a.archive(flat)
			nextStart += a.IntervalMs
		}

		a.current = &Candle{
			BucketStart: bucketStart, Open: price, High: price, Low: price, Close: price,
			Volume: qty, Trades: 1,
		}
		events = append(events, a.makeEvent(CandleEventUpdate, a.current))

	case bucketStart == a.current.BucketStart:
		if price > a.current.High {
			a.current.High = price
		}
		if price < a.current.Low {
			a.current.Low = price
		}
		a.current.Close = price
		a.current.Volume += qty
		a.current.Trades++
		events = append(events, a.makeEvent(CandleEventUpdate, a.current))
	}

	return events
}

func (a *CandleAggregator) Tick(now int64) []CandleEvent {
	if a.current == nil {
		return []CandleEvent{}
	}
	bucketStart := (now / a.IntervalMs) * a.IntervalMs
	if bucketStart <= a.current.BucketStart {
		return []CandleEvent{}
	}
	return a.OnTrade(a.current.Close, 0, bucketStart)
}

func (a *CandleAggregator) Clear() {
	a.current = nil
	a.history = nil
}

func (a *CandleAggregator) Recent(limit int) []Candle {
	extra := 0
	if a.current == nil {
		extra = 1
	}
	start := max(len(a.history)-limit+extra, 0)

	out := make([]Candle, 0, limit)
	for i := start; i < len(a.history); i++ {
		out = append(out, *a.history[i])
	}
	if a.current != nil {
		out = append(out, *a.current)
	}
	if len(out) > limit {
		out = out[len(out)-limit:]
	}
	return out
}

func (a *CandleAggregator) archive(c *Candle) {
	a.history = append(a.history, c)
	if len(a.history) > a.Capacity {
		a.history = a.history[1:]
	}
}

func (a *CandleAggregator) makeEvent(kind CandleEventKind, c *Candle) CandleEvent {
	return CandleEvent{Kind: kind, Interval: a.Interval, Candle: *c}
}
